[Go] Updating a few guidelines and clarifying others (#6903)
Azure/azure-sdk: docs/golang/introduction.md

Your API surface consists of one or more service clients that the consumer instantiates to connect to your service, plus a set of supporting types.

> {% include requirement/MUST id="golang-client-naming-onlyclient" %} name the client `Client`. The combination of the package and the type provide enough context.

Reviewer: Some packages may have multiple clients in them and so we can't name them all Client

}
```

> {% include requirement/SHOULD id="golang-client-naming-onlyclient" %} name the client `Client` if the package name provides enough context.

Reviewer: It really has nothing to do with package name context. The client should be called Client unless you ever expect that more clients will be added to the same package today or tomorrow.

Author: Does this upgrade it to a MUST then?

Reviewer: Yes

Reviewer 2: We should include a blurb about multiple clients in the same package.

Reviewer 3: I prefer we focus on MUST and MUST NOT.  Allows us to be precise in the shape of the SDKs.
`requirement/SHOULD`
Should's create a gray area which leave us without firm ground to argue against.

Author: I've updated the guideline. 

@jhendrixMSFT, I didn't add in a "multiple client" guideline because I'm not sure what we'd say there. It's exceptional enough that I think we'd require a review to figure out the name anyways.

#### Service Client Constructors

> {% include requirement/MUST id="golang-client-constructors" %} provide one or more constructors in the following format that return a new instance of a service client type.  The "simple named" constructor (eg: `NewClient`) MUST use an `azcore.TokenCredential`, assuming the service supports AAD authentication.  If not, then the preferred credential type is used instead.  Constructors MUST return the client instance by reference.

Reviewer: ```suggestion
{% include requirement/MUST id="golang-client-constructors" %} provide one or more constructors in the following format that return a new instance of a service client type.  The "simple named" constructor (eg: `NewClient`) MUST use an `azcore.TokenCredential`, assuming the service supports AAD authentication.  If not, then this constructor should not exist.  Constructors MUST return the client instance by reference.
```

Reviewer: You must always think about the future to avoid breaking changes. A service may not support AAD today, but it might tomorrow. And, in fact, this is fairly common as more services adopt AAD over time. We want to reserve NewClient for AAD ALWAYS.
